eva/storage/sqlite_storage_engine.py

```diff
@@ -0,0 +1,148 @@
+        It dynamically constructs schema in sqlaclchemy
+        to create the table
+        """
+        attr_dict = {"__tablename__": table.name}
+        sqlalchemy_schema = SchemaUtils.get_sqlalchemy_schema(table.columns)
+        attr_dict.update(sqlalchemy_schema)
+        # dynamic schema generation
+        # https://sparrigan.github.io/sql/sqla/2016/01/03/dynamic-tables.html
+        new_table = type("__placeholder_class_name", (BaseModel,), attr_dict)()
+        BaseModel.metadata.tables[table.name].create(self._sql_engine)
+        self._sql_session.commit()
+        return new_table
+
+    def drop(self, table: DataFrameMetadata):
+        try:
+            table_to_remove = BaseModel.metadata.tables[table.name]
+            table_to_remove.drop()
+            self._sql_session.commit()
+            # In-memory metadata does not automatically sync with the database
+            # therefore manually removing the table from the in-memory metadata
+            # https://github.com/sqlalchemy/sqlalchemy/issues/5112
+            BaseModel.metadata.remove(table_to_remove)
+        except Exception as e:
+            logger.exception(
+                f"Failed to drop the table {table.name} with Exception {str(e)}"
+            )
+
+    def write(self, table: DataFrameMetadata, rows: Batch):
+        """
+        Write rows into the sql table.
+
+        Arguments:
+            table: table metadata object to write into
+            rows : batch to be persisted in the storage.
+        """
+        new_table = BaseModel.metadata.tables[table.name]
+        columns = rows.frames.keys()
+        data = []
+        # ToDo: validate the data type before inserting into the table
+        for record in rows.frames.values:
+            row_data = {col: record[idx] for idx, col in enumerate(columns)}
+            data.append(self._dict_to_sql_row(row_data, table.columns))
+        self._sql_engine.execute(new_table.insert(), data)
+        self._sql_session.commit()
+
+    def read(
+        self,
+        table: DataFrameMetadata,
+        batch_mem_size: int,
+    ) -> Iterator[Batch]:
+        """
+        Reads the table and return a batch iterator for the
+        tuples.
+
+        Argument:
+            table: table metadata object of teh table to read
+            batch_mem_size (int): memory size of the batch read from storage
+        Return:
+            Iterator of Batch read.
+        """
+
+        new_table = BaseModel.metadata.tables[table.name]
+        result = self._sql_engine.execute(new_table.select())
+        data_batch = []
+        row_size = None
+        for row in result:
+            # Todo: Verfiy the order of columns in row matches the table.columns
+            # ignore the first dummy (_row_id) primary column
+            data_batch.append(self._sql_row_to_dict(row[1:], table.columns))
+            if row_size is None:
+                row_size = 0
+                row_size = get_size(data_batch)
```

Comment:
get_row_size. Should this be cached in the schema?

Reply:
Yes, we should do it future.